internal/controllers/machinedeployment/machinedeployment_rolling.go

```diff
@@ -74,59 +85,92 @@ func (r *Reconciler) rolloutRolling(ctx context.Context, md *clusterv1.MachineDe
+	// Scale up, if we can.
+	if err := p.reconcileNewMachineSet(ctx); err != nil {
+		return err
+	}
+
+	// Scale down, if we can.
+	return p.reconcileOldMachineSets(ctx)
+}
+
+func (p *rolloutPlanner) reconcileNewMachineSet(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+	allMSs := append(p.oldMSs, p.newMS)
+
+	if *(p.newMS.Spec.Replicas) == *(p.md.Spec.Replicas) {
 		// Scaling not required.
 		return nil
 	}
 
-	if *(newMS.Spec.Replicas) > *(deployment.Spec.Replicas) {
+	if *(p.newMS.Spec.Replicas) > *(p.md.Spec.Replicas) {
 		// Scale down.
-		return r.scaleMachineSet(ctx, newMS, *(deployment.Spec.Replicas), deployment)
+		log.V(5).Info(fmt.Sprintf("Setting scale down intent for %s to %d replicas", p.newMS.Name, *(p.md.Spec.Replicas)), "machineset", client.ObjectKeyFromObject(p.newMS).String())
+		p.scaleIntents[p.newMS.Name] = *(p.md.Spec.Replicas)
+		return nil
 	}
 
-	newReplicasCount, err := mdutil.NewMSNewReplicas(deployment, allMSs, *newMS.Spec.Replicas)
+	newReplicasCount, err := mdutil.NewMSNewReplicas(p.md, allMSs, *p.newMS.Spec.Replicas)
 	if err != nil {
 		return err
 	}
-	return r.scaleMachineSet(ctx, newMS, newReplicasCount, deployment)
-}
-
-func (r *Reconciler) reconcileOldMachineSets(ctx context.Context, allMSs []*clusterv1.MachineSet, oldMSs []*clusterv1.MachineSet, newMS *clusterv1.MachineSet, deployment *clusterv1.MachineDeployment) error {
-	log := ctrl.LoggerFrom(ctx)
 
-	if deployment.Spec.Replicas == nil {
-		return errors.Errorf("spec.replicas for MachineDeployment %v is nil, this is unexpected",
-			client.ObjectKeyFromObject(deployment))
+	if newReplicasCount < *(p.newMS.Spec.Replicas) {
+		scaleDownCount := *(p.newMS.Spec.Replicas) - newReplicasCount
+		log.V(5).Info(fmt.Sprintf("Setting scale down intent for %s to %d replicas (-%d)", p.newMS.Name, newReplicasCount, scaleDownCount), "machineset", client.ObjectKeyFromObject(p.newMS).String())
+		p.scaleIntents[p.newMS.Name] = newReplicasCount
 	}
-
-	if newMS.Spec.Replicas == nil {
-		return errors.Errorf("spec.replicas for MachineSet %v is nil, this is unexpected",
-			client.ObjectKeyFromObject(newMS))
+	if newReplicasCount > *(p.newMS.Spec.Replicas) {
+		scaleUpCount := newReplicasCount - *(p.newMS.Spec.Replicas)
+		log.V(5).Info(fmt.Sprintf("Setting scale up intent for %s to %d replicas (+%d)", p.newMS.Name, newReplicasCount, scaleUpCount), "machineset", client.ObjectKeyFromObject(p.newMS).String())
+		p.scaleIntents[p.newMS.Name] = newReplicasCount
 	}
+	return nil
+}
+
+func (p *rolloutPlanner) reconcileOldMachineSets(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+	allMSs := append(p.oldMSs, p.newMS)
 
-	oldMachinesCount := mdutil.GetReplicaCountForMachineSets(oldMSs)
+	oldMachinesCount := mdutil.GetReplicaCountForMachineSets(p.oldMSs)
 	if oldMachinesCount == 0 {
 		// Can't scale down further
 		return nil
 	}
 
 	allMachinesCount := mdutil.GetReplicaCountForMachineSets(allMSs)
```

Comment:
This is now using a potentially "outdated" newMS.Spec.Replicas. Is this change intended?

Same in l.208

I have some trouble thinking through the consequences of this change

Reply:
```
	newMSReplicas := ptr.Deref(p.newMS.Spec.Replicas, 0)
	if v, ok := p.scaleIntents[p.newMS.Name]; ok {
		newMSReplicas = v
	}
	allMachinesCount := oldMachinesCount + newMSReplicas
```